app/gui2/src/components/CodeEditor.vue

```diff
@@ -0,0 +1,89 @@
+<script setup lang="ts">
+import { onMounted, ref, watch } from 'vue'
+import { useWindowEvent } from '@/util/events'
+import { EditorState } from '@codemirror/state'
+import { EditorView } from '@codemirror/view'
+import { basicSetup } from 'codemirror'
+// y-codemirror.next does not provide type information. See https://github.com/yjs/y-codemirror.next/issues/27
+// @ts-ignore
+import { yCollab } from 'y-codemirror.next'
+import { useGraphStore } from '@/stores/graph'
+import * as Y from 'yjs'
+
+let graphStore = useGraphStore()
+
+// == Keyboard shortcut to toggle the CodeEditor ==
+
+const shown = ref(false)
+const rootElement = ref<HTMLElement>()
+
+useWindowEvent('keydown', (e) => {
+  console.log('keydown', e)
+  const graphEditorInFocus = document.activeElement === document.body
+  const codeEditorInFocus = rootElement.value?.contains(document.activeElement)
+  const validFocus = graphEditorInFocus || codeEditorInFocus
+  const targetKeyPressed = e.key == `\``
+  if (validFocus && targetKeyPressed) {
+    e.preventDefault()
+    shown.value = !shown.value
+  }
+})
+
+// == CodeMirror editor setup  ==
+
+const codeMirrorEl = ref(null)
+const editorView = ref<EditorView>()
+onMounted(() => {
+  watch(
+    () => graphStore.proj.module,
+    (module) => {
+      const yText = module?.contents
+      if (!yText) {
+        console.error('No module content available')
+        return
+      }
+      const undoManager = new Y.UndoManager(yText)
+      const state = EditorState.create({
+        doc: yText.toString(),
+        extensions: [basicSetup, yCollab(yText, null, { undoManager })],
```

Comment:
The `undoManager` and `awareness` already exists inside the `project` store, those should be used here.